Enable scenario test execution in VMR build

mthalman opened the pull request:
Fixes dotnet/source-build#3819
Fixes dotnet/source-build#4278

This enables running the scenario tests via the build scripts.

There will be more follow up work to better integrate the SB smoke tests with the scenario tests as there is some duplication going on there. But for source build scenarios, both test suites are executed.

Both the SB and UB jobs are configured to enable the running of these tests.

It looks like the unified-build has some issues with F# in the OSX jobs that will need to be looked at later.

Test runs (internal links):
* [unified-build](https://dev.azure.com/dnceng/internal/_build/results?buildId=2416222&view=results)
* [source-build](https://dev.azure.com/dnceng/internal/_build/results?buildId=2416085&view=results)

mmitche commented on src/SourceBuild/content/Directory.Build.targets:
```diff
@@ -2,14 +2,18 @@
   <Import Project="Sdk.targets" Sdk="Microsoft.DotNet.Arcade.Sdk" Condition="'$(SkipArcadeSdkImport)' != 'true'" />
 
-  <Target Name="DetermineSourceBuiltSdkVersion">
+  <Target Name="GetDotNetSdkTarball">
```
Nit: `GetDotNetSdkArchive` to be archive extension agnostic.

mthalman replied:
Fixed in https://github.com/dotnet/installer/pull/19222/commits/b74d5fb4c0680ad2022d724d0e00e68585689c19

ViktorHofer commented on src/SourceBuild/content/build.sh:
```diff
@@ -33,12 +34,13 @@ usage()
   echo "  --excludeCIBinarylog            Don't output binary log (short: -nobl)"
   echo "  --prepareMachine                Prepare machine for CI run, clean up processes after build"
   echo "  --use-mono-runtime              Output uses the mono runtime"
+  echo "  --testnobuild                   Run scenario tests without building when invoked with --test"
```
The Arcade extension point for that is the `--projects` flag that you can point in additionally to just build that specific project.

ViktorHofer replied:
https://github.com/dotnet/arcade/blob/986a55a9c7f1ef547bc705d6adada5e0c2217819/eng/common/build.sh#L39

MichaelSimons replied:
> Would that make sense to expose?

I think that would make sense.

mthalman replied:
Fixed in https://github.com/dotnet/installer/pull/19222/commits/81af398edb6090899a4329e158b94b0e04a22937

mthalman commented on src/SourceBuild/content/build.sh:
```diff
@@ -18,7 +18,8 @@ usage()
   echo "  --clean                         Clean the solution"
   echo "  --help                          Print help and exit (short: -h)"
-  echo "  --test                          Run smoke tests (short: -t)"
+  echo "  --test                          Run scenario tests (short: -t)"
```
Fixed in https://github.com/dotnet/installer/pull/19222/commits/aa1f65efd63839fa07607dcdfb722bc19343db02

ViktorHofer commented:
Can we please wait until https://github.com/dotnet/installer/pull/19090 is merged? That's pretty close and it will cause a conflict.

ViktorHofer commented on eng/pipelines/templates/stages/vmr-build.yml:
```diff
@@ -362,6 +364,7 @@ stages:
         container: ${{ variables.androidCrossContainer }}
         targetOS: android
         targetArchitecture: arm64
+        runTests: false
```
False is already the default, no need to mention it.

ViktorHofer replied:
IMHO it bloats the YML. We will define new parameters soon, i.e. `buildPass` which will have a default value of `1` and we won't set that default in every job vertical.

akoeplinger replied:
I agree with Viktor on this one. We don't run tests by default and running tests really only makes sense for full builds, not on verticals, so specifying `runTests: false` on all of these is a lot of bloat. To test the verticals we'll likely need to assemble an sdk+nugets and then run scenario tests against that which is either a separate stage or a separate pipeline.

ViktorHofer commented on src/SourceBuild/content/build.proj:
```diff
@@ -43,6 +53,132 @@
+    <ExtractDotNetSdkArchiveDependsOn Condition="'$(ArchiveExtension)' == '.tar.gz'">$(ExtractDotNetSdkArchiveDependsOn);ExtractDotNetSdkTarball</ExtractDotNetSdkArchiveDependsOn>
+  </PropertyGroup>
+
+  <Target Name="ExtractDotNetSdkArchive"
```
The dotnet sdk layout is already available in the installer repo. Can we just copy that into a well-known location before the repo gets cleaned in installer.proj?

mthalman replied:
I think it's better to consume the actual output rather than bypassing that. It gives better test coverage in case we didn't produce the archive file correctly.

ViktorHofer replied:
I'm fine to keep using the archive but be warry of executable permissions. Anyways, this logic should be moved into installer.proj. I want to keep the build.proj entry point file as minimal as possible. We will also eventually move the Publish target out of build.proj as it doesn't belong there.

mmitche replied:
I don't think this should be in installer.proj either. Perhaps another project altogether?

mmitche replied:
> The dotnet sdk layout is already available in the installer repo. Can we just copy that into a well-known location before the repo gets cleaned in installer.proj?

We should base our testing on the actual produced output, rather than taking a shortcut.

ViktorHofer replied:
> We should base our testing on the actual produced output, rather than taking a shortcut.

Good point, yes that makes sense.

> I don't think this should be in installer.proj either. Perhaps another project altogether?

There are multiple projects that needs the produced and extracted sdk archive: smoke tests, unified build tests, scenario tests and some source-build specific symbol archiving logic. On the other hand, short stack builds don't need it. I think installer.proj is suitable here as it produces that archive and should be responsible for then extracting it and that doesn't run for short stack builds.

mthalman replied:
> source-build specific symbol archiving logic

Can you point to where this uses the extracted SDK?

> I think installer.proj is suitable here as it produces that archive and should be responsible for then extracting it and that doesn't run for short stack builds.

In terms of separation of concerns, it seems to me that extracting the SDK is not part of the build. It's only a necessary step because of testing. So doing it by default in a build is unnecessary and should be left to the test logic.

ViktorHofer replied:
If extracting the SDK would only be necessary for testing, then it shouldn't be done as part of the build, correct. But that's not the case as creating the sdk symbol archiving already needs the extracted layout: https://github.com/dotnet/installer/blob/144a405fd5fe407d34cf86ed4e68ec34268eca5b/src/SourceBuild/content/eng/build.sourcebuild.targets#L46

ViktorHofer commented on src/SourceBuild/content/eng/build.sourcebuild.targets:
```diff
@@ -1,7 +1,9 @@
   <PropertyGroup>
-    <SmokeTestsDir>$([MSBuild]::NormalizeDirectory('$(RepoRoot)', 'test', 'Microsoft.DotNet.SourceBuild.SmokeTests'))</SmokeTestsDir>
+    <SmokeTestProjectName>Microsoft.DotNet.SourceBuild.SmokeTests</SmokeTestProjectName>
+    <SmokeTestsDir>$([MSBuild]::NormalizeDirectory('$(RepoRoot)', 'test', '$(SmokeTestProjectName)'))</SmokeTestsDir>
```
Now that we will add more tests under `~/tests` it would make sense to add a property to the root Directory.Build.props for the tests directory.

mthalman replied:
This is made obsolete by https://github.com/dotnet/installer/pull/19290

ViktorHofer commented on src/SourceBuild/content/build.sh:
```diff
@@ -180,7 +182,15 @@ while [[ $# > 0 ]]; do
+    -testnobuild)
+      test_no_build=true
+      ;;
+    -test-projects)
+      # In order to support multiple values delimited by semicolons, it needs to be properly quoted so that
+      # MSBuild can parse it correctly. See https://github.com/dotnet/sdk/issues/8792#issuecomment-393756980
+      properties="$properties /p:DotNetTestProjectsToRun='\"$2\"'"
+      shift
+      ;;
```
I would rather use the official Arcade extension point which is already well known by repo contributors. Is that possible? https://github.com/dotnet/arcade/blob/0ffb8e9f4a064d8c77f07420dc399baec9daef89/eng/common/build.ps1#L68

ViktorHofer replied:
The `Projects` / `-projects` property/switch signals which projects to pass in to the Build.proj / build wrapper. The action that is performed on that project depends on the action being passed in, i.e. `-test`, `-restore`, `-build`, etc. This switch would make it easier to build individual task projects or build and test test projects inside the VMR.

I don't think this is beyond the scope of this PR as I disagree with the `-test-projects` switch. It should just be `-projects` to be consistent with all other Arcade-ified repos.

mthalman replied:
Ok, it'll be consistent in name then but not in functionality (because it won't apply to build), so how does that help?

mthalman replied:
I've reverted the changes to support this: https://github.com/dotnet/installer/pull/19222/commits/28ddf277572e32ae696a73f2dd4c2d38b714040f

ViktorHofer commented on src/SourceBuild/content/eng/build.ps1:
```diff
@@ -66,6 +74,19 @@ function Build {
     @properties
 }
 
+function Test {
+  InitializeToolset
+
+  $bl = if ($binaryLog) { '/bl:' + (Join-Path $LogDir 'ScenarioTests.binlog') } else { '' }
+  $tp = if ($testProjects) { '/p:DotNetTestProjectsToRun="' + $testProjects + '"' } else { '' }
+
+  MSBuild $buildProj `
+    $bl `
+    /t:Test `
+    $tp `
+    @properties
+}
+
```
Same comment as in build.sh. We shouldn't need a separate function and MSBuild invocation and instead just pass the Test property in same as Arcade's build scripts.

mthalman replied:
These changes have been removed in https://github.com/dotnet/installer/pull/19222/commits/49c04fc531de986bab872236a4cb7a8b651be539 as part of integrating with the new test infra from https://github.com/dotnet/installer/pull/19290.

ViktorHofer commented on src/SourceBuild/content/Directory.Build.targets:
```diff
@@ -2,14 +2,18 @@
     </ItemGroup>
+  </Target>
 
+  <Target Name="DetermineSourceBuiltSdkVersion"
+          DependsOnTargets="GetDotNetSdkArchive">
```
Why do we need a separate target for this?

mmitche replied:
While I think using the archive file name is fine, we could consider another method for extracting versions if we find we keep doing this kind of thing. We could take the vertical manifest and find a package with a matching version. This is how Maestro dependency flow typically works for assets. If installer wants to download aspnetcore's runtime archive, it will find an aspnet runtime package that will have a matching version, use that in its Version.Details.xml, then construct the file name off the version property.

The SDK doesn't produce a "sentinel" package today in the VMR. The regular official build does (`Microsoft.Dotnet.Sdk.Internal`).

ViktorHofer commented on eng/pipelines/templates/jobs/vmr-build.yml:
```diff
@@ -440,17 +458,30 @@ jobs:
   # Only upload test results if enabled
   - ${{ if eq(parameters.runTests, 'True') }}:
+    - ${{ if eq(parameters.buildSourceOnly, 'True') }}:
+      - task: PublishTestResults@2
+        displayName: Publish Source Build Smoke Test Results
+        condition: succeededOrFailed()
+        continueOnError: true
+        inputs:
+          testRunner: vSTest
+          testResultsFiles: 'test/**/*.trx'
+          searchFolder: $(sourcesPath)
+          mergeTestResults: true
+          publishRunAttachments: true
+          testRunTitle: SourceBuild_SmokeTests_$(Agent.JobName)
```
Could we instead tell the smoke tests project to put the test result files under the known artifacts/TestResults/ location?

mthalman replied:
This is now implemented by https://github.com/dotnet/installer/pull/19290

ViktorHofer commented on src/SourceBuild/content/build.proj:
```diff
@@ -43,6 +53,132 @@
+
+  <Target Name="Test" DependsOnTargets="ExtractDotNetSdkArchive;ExtractScenarioTestsPackage;$(RunTestsTargets)" />
+
+  <Target Name="RunScenarioTests"
+          Condition="'$(DotNetTestProjectsToRun)' == '' or $(DotNetTestProjectsToRun.Contains('$(ScenarioTestsProjectName)'))">
+    <Message Text="Running scenario tests: $(TargetArchitecture) $(Configuration) $(TargetOS) $(TargetRid)" Importance="high" />
+
+    <ItemGroup>
+      <_DotNetPath Include="$(DotNetInstallDir)sdk/**/dotnet.dll" />
+    </ItemGroup>
+
+    <PropertyGroup>
+      <_CurrentDateTime>$([System.DateTime]::Now.ToString("yyyy-MM-dd_HH_mm_ss"))</_CurrentDateTime>
+      <_TestXmlOutputPath>$(ScenarioTestsResultsDir)$(_CurrentDateTime).xml</_TestXmlOutputPath>
+      <_ScenarioTestsArgs>--dotnet-root $(DotNetInstallDir) --xml $(_TestXmlOutputPath) --target-rid $(TargetRid) --test-root $(ScenarioTestsRunsDir)$(_CurrentDateTime) --no-cleanup --no-traits Category=MultiTFM</_ScenarioTestsArgs>
+      <_MSBuildSdksDir>/%(_DotNetPath.Directory)Sdks</_MSBuildSdksDir>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <_ScenarioTestsEnvVars Include="LOCAL_FEED_PATH=$(ArtifactsShippingPackagesDir)" />
+      <_ScenarioTestsEnvVars Include="MSBuildSDKsPath=$(_MSBuildSdksDir)" />
+    </ItemGroup>
+
+    <MakeDir Directories="$(ScenarioTestsResultsDir);$(ScenarioTestsRunsDir)" />
+
+    <Exec Command="$(DotnetTool) @(_ScenarioTestDll) $(_ScenarioTestsArgs)"
+          EnvironmentVariables="@(_ScenarioTestsEnvVars)" />
+  </Target>
```
This belongs into the scenario tests project and not into build.proj. Otherwise adding a new test project wouldn't scale well.

ViktorHofer commented on src/SourceBuild/content/build.proj:
```diff
@@ -43,6 +53,132 @@
+    <_InstallerNuGetConfig>$(RepoRoot)src/installer/NuGet.config</_InstallerNuGetConfig>
+  </PropertyGroup>
+
+  <Target Name="SetupNuGetConfig"
+          Inputs="$(_InstallerNuGetConfig)"
+          Outputs="$(_ScenarioTestsNuGetConfig)">
+
+    <Copy SourceFiles="$(_InstallerNuGetConfig)"
+          DestinationFiles="$(_ScenarioTestsNuGetConfig)" />
+
+    <AddSourceToNuGetConfig NuGetConfigFile="$(_ScenarioTestsNuGetConfig)"
+                            SourceName="local-feed"
+                            SourcePath="$(ScenarioTestsLocalFeedDir)" />
+  </Target>
+
+  <ItemGroup>
+    <_ArtifactsPackage Include="$(ArtifactsShippingPackagesDir)**;
+                                $(ArtifactsNonShippingPackagesDir)**"
+                       Condition="!$([System.String]::Copy('%(Identity)').EndsWith('.symbols.nupkg'))" />
+  </ItemGroup>
+
+  <Target Name="CreateScenarioTestsLocalFeed">
+
+    <RemoveDir Directories="$(ScenarioTestsLocalFeedDir)" />
+    <MakeDir Directories="$(ScenarioTestsLocalFeedDir)" />
+    <!-- Combine all packages into a single directory with symlinks so that we can have a single feed that points to them all. -->
+    <Copy SourceFiles="@(_ArtifactsPackage)"
+          DestinationFolder="$(ScenarioTestsLocalFeedDir)"
+          UseSymbolicLinksIfPossible="true" />
+
+  </Target>
```
Why can't scenario-tests use the same infrastructure to flow dependencies across inner-repositories?

mmitche replied:
Assuming we don't need non-shipping packages for testing, I don't think the repo dependency flow structure applies here. It would also be awkward, since you'd have to introduce all the dependencies into the V.D.xml file, when the repo build wouldn't actually depend on them to build (there would be no use of the version properties).

ViktorHofer replied:
Scenario-tests should be able to consume the live built non-shipping packages, i.e. Microsoft.DotNet.XUnitExtensions which we want to use there.

mmitche replied:
> How will scenario-tests know which version of runtime packages to reference when building the tests?

The test infra has a nuget feed config like any other, and the tests themselves (typically templatles) are doing sdk operations. The SDK is the one defining what versions should be used.

MichaelSimons commented:
I was playing around with the dev UX a bit and was surprised at the overhead prior to when it appears to actually build the test projects in a VMR that was previously built.  Looking through the output it looks like it is creating the package archives every time I run ./build.sh --test -sb.  It doesn't seem right that this logic is running in this scenario.

```
  Packaged all symbols in '/workspaces/dotnet/artifacts/assets/Release/dotnet-symbols-all-9.0.100-preview.4.24202.1-fedora.39-x64.tar.gz'
  Packaged sdk symbols in '/workspaces/dotnet/artifacts/assets/Release/dotnet-symbols-sdk-9.0.100-preview.4.24202.1-fedora.39-x64.tar.gz'
  Found 0 files in prebuilt packages dir.
  Shipping packages are located in '/workspaces/dotnet/artifacts/packages/Release/Shipping/'.
  Shipping assets are located in '/workspaces/dotnet/artifacts/assets/Release/'.
```

MichaelSimons commented on src/SourceBuild/content/build.sh:
```diff
@@ -33,12 +34,14 @@ usage()
   echo "  --prepareMachine                Prepare machine for CI run, clean up processes after build"
   echo "  --use-mono-runtime              Output uses the mono runtime"
+  echo "  --testnobuild                   Run tests without building when invoked with --test"
+  echo "  --test-projects <NAMES>         Run tests for the specified projects"
```
The name and description can give the impression that the `--test` option is not necessary.

mthalman replied:
I've reverted the changes to support this: https://github.com/dotnet/installer/commit/28ddf277572e32ae696a73f2dd4c2d38b714040f

mthalman commented:
> I was playing around with the dev UX a bit and was surprised at the overhead prior to when it appears to actually build the test projects in a VMR that was previously built. Looking through the output it looks like it is creating the package archives every time I run ./build.sh --test -sb. It doesn't seem right that this logic is running in this scenario.
> 
> ```
>   Packaged all symbols in '/workspaces/dotnet/artifacts/assets/Release/dotnet-symbols-all-9.0.100-preview.4.24202.1-fedora.39-x64.tar.gz'
>   Packaged sdk symbols in '/workspaces/dotnet/artifacts/assets/Release/dotnet-symbols-sdk-9.0.100-preview.4.24202.1-fedora.39-x64.tar.gz'
>   Found 0 files in prebuilt packages dir.
>   Shipping packages are located in '/workspaces/dotnet/artifacts/packages/Release/Shipping/'.
>   Shipping assets are located in '/workspaces/dotnet/artifacts/assets/Release/'.
> ```

This seems outside the context of testing. Does this happen when you don't use the `--test` option? If so, that's a separate issue.

MichaelSimons commented:
> > I was playing around with the dev UX a bit and was surprised at the overhead prior to when it appears to actually build the test projects in a VMR that was previously built. Looking through the output it looks like it is creating the package archives every time I run ./build.sh --test -sb. It doesn't seem right that this logic is running in this scenario.
> > ```
> >   Packaged all symbols in '/workspaces/dotnet/artifacts/assets/Release/dotnet-symbols-all-9.0.100-preview.4.24202.1-fedora.39-x64.tar.gz'
> >   Packaged sdk symbols in '/workspaces/dotnet/artifacts/assets/Release/dotnet-symbols-sdk-9.0.100-preview.4.24202.1-fedora.39-x64.tar.gz'
> >   Found 0 files in prebuilt packages dir.
> >   Shipping packages are located in '/workspaces/dotnet/artifacts/packages/Release/Shipping/'.
> >   Shipping assets are located in '/workspaces/dotnet/artifacts/assets/Release/'.
> > ```
> 
> This seems outside the context of testing. Does this happen when you don't use the `--test` option? If so, that's a separate issue.

It seems there are two issues here.  One, why is build even happening - is there not a way to avoid rebuilding?  This feels related to arcade not being used.  w/arcade --build is separate from --test so you have the ability to do both together or independently.  Two, the symbols should not be repackaged when the build outputs haven't changed.  I agree the later is outside the scope here.

mthalman commented:
>  One, why is build even happening - is there not a way to avoid rebuilding?

That's what the `--testnobuild` option is for.

akoeplinger commented on eng/pipelines/templates/jobs/vmr-build.yml:
```diff
@@ -440,17 +458,30 @@ jobs:
+        condition: succeededOrFailed()
+        continueOnError: true
+        inputs:
+          testRunner: vSTest
```
```suggestion
          testRunner: VSTest
```

mmitche commented on src/SourceBuild/content/build.proj:
```diff
@@ -43,6 +53,132 @@
+
+  <ItemGroup>
+    <_ArtifactsPackage Include="$(ArtifactsShippingPackagesDir)**;
+                                $(ArtifactsNonShippingPackagesDir)**"
```
My hunch is that non-shipping is not required for running the scenario tests.

mthalman replied:
Fixed in https://github.com/dotnet/installer/commit/6450dbc9202d660547c82880df49100c3a8a07bc

mmitche commented on src/SourceBuild/content/build.proj:
```diff
@@ -43,6 +53,132 @@
+          Outputs="@(_DirectoryProjFiles)">
+    <!-- Define Directory.Build.* files in the scenario tests artifacts directory to prevent the test run from
+         picking up the configuration from the VMR's Directory.Build.* files. We need an isolated configuration
+         from which to test. -->
```
Does MSBuild automatically import `Directory.Build.*` from directories above?

mthalman replied:
MSBuild will scan the directory hierarchy, starting from the project that's being built and moving up through the parents until it finds `Directory.Build.*`. At that point it will stop. We need to prevent it from getting the VMR's definition of this file so this creates stub files which will prevent MSBuild from looking further.

mmitche commented on src/SourceBuild/content/build.proj:
```diff
@@ -43,6 +53,132 @@
+    <Unzip SourceFiles="@(SdkArchiveItem)" DestinationFolder="$(DotNetInstallDir)" />
+  </Target>
+
+  <Target Name="ExtractScenarioTestsPackage">
```
nit
```suggestion
  <Target Name="ExtractScenarioTestsPackages">
```

mthalman replied:
Fixed in https://github.com/dotnet/installer/pull/19222/commits/4ad40634d16a550ede3a66ebd4a058702e5facb7

mmitche commented on src/SourceBuild/content/build.sh:
```diff
@@ -33,12 +34,14 @@ usage()
   echo "  --excludeCIBinarylog            Don't output binary log (short: -nobl)"
   echo "  --prepareMachine                Prepare machine for CI run, clean up processes after build"
   echo "  --use-mono-runtime              Output uses the mono runtime"
+  echo "  --testnobuild                   Run tests without building when invoked with --test"
```
IMO `--test-no-build` is more readable.

ViktorHofer replied:
FWIW runtime has `testnobuild`: https://github.com/dotnet/runtime/blob/5c3bfeb8477b34a15ef851faa54cae1968ab4237/eng/build.sh#L72

mthalman replied:
This is now implemented from https://github.com/dotnet/installer/pull/19290

mmitche commented on src/SourceBuild/content/build.sh:
```diff
@@ -193,49 +203,57 @@ if [[ "$ci" == true ]]; then
+    MSBuild "$projectPath" \
+      -t:$target \
       $bl \
       /p:Configuration=$configuration \
```
Can Configuration be added as part of `$properties` rather than passed separately?

mthalman replied:
In the interest of limiting the changes in this PR, I'm going to decline making this change since this implementation already existed.

ViktorHofer commented on src/SourceBuild/content/build.proj:
```diff
@@ -43,6 +53,132 @@
+
+  <Target Name="Test" DependsOnTargets="ExtractDotNetSdkArchive;ExtractScenarioTestsPackage;$(RunTestsTargets)" />
+
+  <Target Name="RunScenarioTests"
+          Condition="'$(DotNetTestProjectsToRun)' == '' or $(DotNetTestProjectsToRun.Contains('$(ScenarioTestsProjectName)'))">
+    <Message Text="Running scenario tests: $(TargetArchitecture) $(Configuration) $(TargetOS) $(TargetRid)" Importance="high" />
+
+    <ItemGroup>
+      <_DotNetPath Include="$(DotNetInstallDir)sdk/**/dotnet.dll" />
+    </ItemGroup>
+
+    <PropertyGroup>
+      <_CurrentDateTime>$([System.DateTime]::Now.ToString("yyyy-MM-dd_HH_mm_ss"))</_CurrentDateTime>
+      <_TestXmlOutputPath>$(ScenarioTestsResultsDir)$(_CurrentDateTime).xml</_TestXmlOutputPath>
+      <_ScenarioTestsArgs>--dotnet-root $(DotNetInstallDir) --xml $(_TestXmlOutputPath) --target-rid $(TargetRid) --test-root $(ScenarioTestsRunsDir)$(_CurrentDateTime) --no-cleanup --no-traits Category=MultiTFM</_ScenarioTestsArgs>
+      <_MSBuildSdksDir>/%(_DotNetPath.Directory)Sdks</_MSBuildSdksDir>
+    </PropertyGroup>
+
+    <ItemGroup>
+      <_ScenarioTestsEnvVars Include="LOCAL_FEED_PATH=$(ArtifactsShippingPackagesDir)" />
+      <_ScenarioTestsEnvVars Include="MSBuildSDKsPath=$(_MSBuildSdksDir)" />
+    </ItemGroup>
+
+    <MakeDir Directories="$(ScenarioTestsResultsDir);$(ScenarioTestsRunsDir)" />
+
+    <Exec Command="$(DotnetTool) @(_ScenarioTestDll) $(_ScenarioTestsArgs)"
+          EnvironmentVariables="@(_ScenarioTestsEnvVars)" />
+  </Target>
```
Why are we running scenario-tests here and not from within the scenario-tests repo?

mthalman replied:
Fixed as part of https://github.com/dotnet/installer/pull/19222/commits/7bef2e3fcfe6bf3763bc74e869529a080ed3ea3a and https://github.com/dotnet/scenario-tests/pull/85

mthalman commented on src/SourceBuild/content/build.proj:
```diff
@@ -43,6 +53,132 @@
+    <_InstallerNuGetConfig>$(RepoRoot)src/installer/NuGet.config</_InstallerNuGetConfig>
+  </PropertyGroup>
+
+  <Target Name="SetupNuGetConfig"
```
For SB, we don't want to use the feed of packages that were just built. That's not how customers would use a source built SDK.

mthalman commented on src/SourceBuild/content/build.sh:
```diff
@@ -367,4 +385,11 @@ if [[ "$sourceOnly" == "true" ]]; then
+
+if [ "$test" == "true" ]; then
+  Test
+fi
```
These changes have been removed in https://github.com/dotnet/installer/pull/19222/commits/2621b96ed1f9077757b089187b767b87e23a136f as part of integrating with the new test infra from https://github.com/dotnet/installer/pull/19290.

mthalman commented:
I've addressed all comments.

There was also a bunch of changes needed to integrate with the changes from https://github.com/dotnet/installer/pull/19290. This is mostly done in https://github.com/dotnet/installer/pull/19222/commits/2621b96ed1f9077757b089187b767b87e23a136f. That removes the explicit invocation of the scenario tests executable from build.proj and pushes things into tests.proj. Rather than explicitly running the scenario tests executable, it is now integrated with Arcade's test execution just as is done with the other test projects. There is a difference with those, however, because the other ones can use the `VSTest` target which doesn't apply to the scenario tests. So it calls it with the `Test` target. But it also needs some custom configuration to setup the enviroment it'll run in. Anyway, I think I've got an implementation that makes things clean from the calling side.

mmitche commented on src/SourceBuild/content/repo-projects/installer.proj:
```diff
@@ -78,11 +78,20 @@
           UseHardlinksIfPossible="true" />
+
+    <!-- Extract the SDK archive so its contents can be used downstream -->
+    <MakeDir Directories="$(DotNetSdkExtractDir)" />
```
Is this always used? (symbol layout and tests). Maybe conditionalize the extraction it if it won't be used later.

ViktorHofer replied:
This is a product dependency for source-build as part of the finish-source-only.proj and ideally also a test dependency by the two test projects. Today those test projects extract the SDK archive yet again and it would be good to consolidate that in a follow-up.

I agree that this should be conditioned to only run when needed. With the new pattern of encapsulating code units we can achieve that and declare the dependency directly.

Meaning, I would put this into a eng/extract-sdk-archive.proj NoTargets msbuild sdk file and then reference that from finish-source-only.proj and the test projects. That's what we do with publish.proj which creates the merged asset manifest.

mthalman replied:
Fixed in https://github.com/dotnet/installer/pull/19222/commits/41b46b4cc39f962914c83d2257e9a2f4b75364e3

mmitche commented on src/SourceBuild/content/test/scenario-tests.proj:
```diff
@@ -0,0 +1,53 @@
+<Project Sdk="Microsoft.Build.Traversal">
```
So is the intention here to treat the scenario tests repo as "part" of the VMR, rather than a separate component that is lives in the VMR, like runtime, installer, etc.

ViktorHofer replied:
I want to warn on directly referencing projects from an inner repository. `ProjectReference` items transfer additional state. Here this means that additional state from the VMR build infrastructure might be passed into the scenario-tests projects within the scenario-tests repo.

This might not be an issue today but this crosses the repository boundary which I would prefer to avoid when possible. When we discussed this a few weeks ago, I was under the assumption that we could just call into the scenario-tests build entrypoint and pass the `-test` action and additional msbuild properties in by leveraging the already existing `repo-projects/scenario-tests.proj` project.

mthalman replied:
Fixed in https://github.com/dotnet/installer/pull/19222/commits/09b93d4d4ac9a4d396086c5e96dc7b15a52359d8

ViktorHofer commented on eng/pipelines/templates/jobs/vmr-build.yml:
```diff
@@ -460,18 +460,43 @@ jobs:
 
   # Only upload test results if enabled
   - ${{ if eq(parameters.runTests, 'True') }}:
+    - ${{ if eq(parameters.buildSourceOnly, 'True') }}:
+      - task: PublishTestResults@2
+        displayName: Publish Source Build Smoke Test Results
+        condition: succeededOrFailed()
+        continueOnError: true
+        inputs:
+          testRunner: VSTest
+          testResultsFiles: 'test/**/*.trx'
+          searchFolder: $(sourcesPath)
+          mergeTestResults: true
+          publishRunAttachments: true
+          testRunTitle: SourceBuild_SmokeTests_$(Agent.JobName)
+
```
This block should be removed. Smoke tests don't publish test results into the test folder anymore.

mthalman replied:
Fixed in https://github.com/dotnet/installer/pull/19222/commits/8e9b075f98eb1bb5ec618d1964b8bf679c271d71

ViktorHofer commented on eng/pipelines/templates/jobs/vmr-build.yml:
```diff
@@ -460,18 +460,43 @@ jobs:
+      continueOnError: true
+      inputs:
+        testRunner: xUnit
+        testResultsFiles: 'artifacts/TestResults/**/*.xml'
```
There might be other xml files in that directory already today or eventually in the future that aren't tied to scenario tests. It think it would make sense to introduce a subfolder for scenario-tests, i.e. `artifacts/TestResults/Release/scenario-tests`.

MichaelSimons replied:
Agreed that the tests results should be separated.  This includes the artifacts gathered in `Prepare BuildLogs staging directory and check assets`

mthalman replied:
Fixed in https://github.com/dotnet/installer/pull/19222/commits/7bb0e4a3be844601edee86c6f66d76a702a8a056

ViktorHofer commented on src/SourceBuild/content/test/tests.proj:
```diff
@@ -7,4 +7,49 @@
 
-</Project>
\ No newline at end of file
+  <Target Name="RunCoreTests">
+    <MSBuild Projects="$(MSBuildThisFileFullPath)"
+             Properties="Configuration=$(Configuration)"
+             ContinueOnError="true"
+             Targets="Build;VSTest" />
+
+    <PropertyGroup>
+      <CoreTestsResult>$(MSBuildLastTaskResult)</CoreTestsResult>
+    </PropertyGroup>
+  </Target>
```
We shouldn't need a separate target for this. MSBuild can invoke multiple entry point targets: `VSTest;RunScenarioTests`.

mthalman replied:
The problem with this is that any failure in the `VSTest` target will stop execution, preventing the `RunScenarioTests` target from running. So you don't get a full view of all test results until you can get the first target to be successful.

mthalman replied:
> This guarantees that when I do `dotnet test test/tests.proj`, we use the same same entry point as when doing `build.cmd -test`.

Why is attempting to support `dotnet test` even a scenario? There's a whole bunch of MSBuild properties that need to be set in order for the tests to do the right thing. This is all done by the build script.

ViktorHofer replied:
We use the Traversal msbuild sdk which is quite flexible. Does one of these setting help? https://github.com/microsoft/MSBuildSdks/blob/32403937dc2bb90f74c2156b9e140538b3f25d51/src/Traversal/Sdk/Traversal.targets#L210-L211

ViktorHofer replied:
I quickly tried this out:

```xml
<Project Sdk="Microsoft.NET.Sdk">

  <PropertyGroup>
    <OutputType>Exe</OutputType>
    <TargetFramework>net9.0</TargetFramework>
    <ImplicitUsings>enable</ImplicitUsings>
    <Nullable>enable</Nullable>
  </PropertyGroup>

  <Target Name="A">
    <Error Text="A target" ContinueOnError="WarnAndContinue" />
  </Target>

  <Target Name="B">
    <Error Text="B target" />
  </Target>

</Project>
```

> PS C:\temp\multitarg> dotnet msbuild -t:"A;B"
MSBuild version 17.10.0-preview-24101-01+07fd5d51f for .NET
  multitarg succeeded with warnings (0.0s)
    C:\temp\multitarg\multitarg.csproj(11,5): warning : A target [C:\temp\multitarg\multitarg.csproj]
  multitarg failed with errors (0.0s)
    C:\temp\multitarg\multitarg.csproj(15,5): error : B target [C:\temp\multitarg\multitarg.csproj]

---

That said, I think it would make more sense to just define the Test/VSTest target in scenario-tests.proj and just let Traversal SDK invoke that target as it already does with the other two test projects. In my head tests.proj could just look like this:

```xml
<Project Sdk="Microsoft.Build.Traversal">

  <ItemGroup>
    <ProjectReference Include="Microsoft.DotNet.SourceBuild.SmokeTests\Microsoft.DotNet.SourceBuild.SmokeTests.csproj"
                      Condition="'$(DotNetBuildSourceOnly)' == 'true'" />
    <ProjectReference Include="Microsoft.DotNet.UnifiedBuild.Tests\Microsoft.DotNet.UnifiedBuild.Tests.csproj"
                      Condition="'$(ShortStack)' != 'true' and '$(PortableBuild)' == 'true' and '$(PgoInstrument)' != 'true'" />
    <ProjectReference Include="scenario-tests.proj" />
  </ItemGroup>

</Project>
```

This would have the benefit of keeping a single entry point target and reduce complexity from tests.proj and move that into scenario-tests.proj itself.

mthalman replied:
> I think it would make more sense to just define the VSTest target in scenario-tests.proj and just let Traversal SDK invoke that target as it already does with the other two test projects.

Unfortunately, that approach doesn't work. I can't override the `VSTest` target. It ends up getting overridden because `Microsoft.TestPlatform.ImportAfter.targets` and `Microsoft.TestPlatform.targets` get imported after the definition of scenario-tests.proj.

mthalman replied:
> ```xml
>   <Target Name="A">
>     <Error Text="A target" ContinueOnError="WarnAndContinue" />
>   </Target>
> 
>   <Target Name="B">
>     <Error Text="B target" />
>   </Target>
> ```

This isn't really possible without introducing a new target. I can't influence the `VSTest` target to have the behavior that you've shown in target `A`. This is what my original implementation solves.

mthalman replied:
Fixed in https://github.com/dotnet/installer/pull/19222/commits/1c9cdd7248d20ace14d028008d1edf7bb395ba0f

ViktorHofer commented on src/SourceBuild/content/test/tests.proj:
```diff
@@ -7,4 +7,49 @@
+    </PropertyGroup>
+  </Target>
+
+  <Target Name="RunTests" DependsOnTargets="RunCoreTests;RunScenarioTests">
+    <Error Condition="'$(CoreTestsResult)' == 'false' or '$(ScenarioTestsResult)' == 'false'" Text="Test failures occurred." />
+  </Target>
```
Same comment as above, we shouldn't need this target and should call the `VSTest` and the `RunScenarioTests` targets directly.

mthalman replied:
Fixed in https://github.com/dotnet/installer/pull/19222/commits/1c9cdd7248d20ace14d028008d1edf7bb395ba0f

ViktorHofer commented on src/SourceBuild/content/test/tests.proj:
```diff
@@ -7,4 +7,49 @@
+       some setup steps to be configured with the VMR. -->
+  <Target Name="RunScenarioTests">
+    <PropertyGroup>
+      <_HostArchitecture>$([System.Runtime.InteropServices.RuntimeInformation]::OSArchitecture)</_HostArchitecture>
+      <_RunScenarioTests>true</_RunScenarioTests>
+
+      <!-- Skip scenario tests if the host architecture is different from the target architecture since the tests
+           require the ability to execute the built SDK. But the CLI is not capable of running on a host with a 
+           different architecture (i.e. "cannot execute binary file: Exec format error"). -->
+      <_RunScenarioTests Condition="'$(_HostArchitecture.ToLower())' != '$(TargetArchitecture.ToLower())'">false</_RunScenarioTests>
```
Can't this be expressed via the already defined variables in Directory.Build.props?

I.e.
```xml
<_RunScenarioTests Condition="'$(BuildArchitecture)' != '$(TargetArchitecture)''">false</_RunScenarioTests>
```

mthalman replied:
Fixed in https://github.com/dotnet/installer/pull/19222/commits/b7bf1bb33102e085eac8e96121d081e3c288a32b

ViktorHofer reviewed (commented):
I just noticed that scenario-tests aren't build or run on Windows, presumably because of the `__PortableTargetOS` condition. That variable is only set in build.sh.

mthalman commented on src/SourceBuild/content/Directory.Build.targets:
```diff
@@ -2,15 +2,22 @@
   <Import Project="Sdk.targets" Sdk="Microsoft.DotNet.Arcade.Sdk" Condition="'$(SkipArcadeSdkImport)' != 'true'" />
 
-  <Target Name="DetermineSourceBuiltSdkVersion">
+  <Target Name="GetDotNetSdkArchive">
```
Reverted these changes in https://github.com/dotnet/installer/pull/19222/commits/fa6b2e41ab74487c1ce5d88c8528d83bdb97ac75.

mthalman commented:
> I just noticed that scenario-tests aren't build or run on Windows, presumably because of the `__PortableTargetOS` condition. That variable is only set in build.sh.

Fixed in https://github.com/dotnet/installer/pull/19222/commits/9745f7ad22ade03719c6882e91d31ba4070078bf

mthalman commented:
There are issues getting the scenario tests working with the DevVersions leg for the UB pipeline. There are a few issues with it:
* It requires the use of the `--fx-version` in the `dotnet` command when running the test executable. This is because the scenario-tests are built with the SDK installed by Arcade (with an official product version). The runtime version it is associated with is not compatible with the runtime associated with the Dev-built SDK. You can override this with the `--fx-version` option. This is shown in https://github.com/dotnet/installer/pull/19222/commits/d14add1f6aba3265d03e70ada527840c7b73a3bc.
* The other bigger problem is with packages that get restored as part of building the projects that are tested by the scenario tests. This would require restoring from the locally built packages. But even that isn't enough because it attempts to load workload packages as well which haven't been locally built. And there are no versions that exist that are associated with the Dev version of the SDK that was built. Changes that attempted to get this to work: https://github.com/dotnet/installer/pull/19222/commits/cb6d71abace7bb4a3ac3eb5890ae16181e69b5b7.

Both of these commits were reverted. This could potentially be made to work (such as filtering out tests that require this) but would require further changes that I don't want to get into in this PR at this time.

mthalman commented:
Note to reviewers:
I really need to wrap up this PR so I can focus on other work. If you identify functional issues with these changes, I will continue to fix them. But if they are non-functional/refactoring types of changes or expanding the scope, I will log issues to have that addressed later.

ViktorHofer commented on eng/pipelines/templates/jobs/vmr-build.yml:
```diff
@@ -474,18 +478,31 @@ jobs:
 
   # Only upload test results if enabled
   - ${{ if eq(parameters.runTests, 'True') }}:
+
```
nit:

```suggestion
```

ViktorHofer commented on src/SourceBuild/content/eng/build.ps1:
```diff
@@ -59,7 +59,7 @@ $targets = "/t:Build"
 if ($test) {
   $project = Join-Path (Join-Path $RepoRoot "test") "tests.proj"
-  $targets += ";VSTest"
+  $targets = "$targets;VSTest"
```
undo this chnage:

```suggestion
  $targets += ";$VSTest"
```

ViktorHofer commented on src/SourceBuild/content/eng/extract-sdk-archive.proj:
```diff
@@ -0,0 +1,18 @@
+    <TargetFramework>$(NetCurrent)</TargetFramework>
+  </PropertyGroup>
+
+  <Target Name="ExtractSdkArchive" BeforeTargets="AfterBuild" DependsOnTargets="DetermineSourceBuiltSdkVersion">
```
Inputs & Outputs for incremental build support:

```suggestion
  <Target Name="ExtractSdkArchive"
          BeforeTargets="AfterBuild"
          DependsOnTargets="DetermineSourceBuiltSdkVersion"
          Inputs="$(SdkTarballPath)"
          Outputs="$(DotNetSdkExtractDir)">
```

ViktorHofer commented on src/SourceBuild/content/eng/extract-sdk-archive.proj:
```diff
@@ -0,0 +1,18 @@
+    <Unzip Condition="'$(ArchiveExtension)' == '.zip'"
+           SourceFiles="$(SdkTarballPath)"
+           DestinationFolder="$(DotNetSdkExtractDir)" />
+
```
```suggestion
```

ViktorHofer commented on src/SourceBuild/content/eng/extract-sdk-archive.proj:
```diff
@@ -0,0 +1,18 @@
+  </PropertyGroup>
+
+  <Target Name="ExtractSdkArchive" BeforeTargets="AfterBuild" DependsOnTargets="DetermineSourceBuiltSdkVersion">
+
```
```suggestion
```

ViktorHofer commented on src/SourceBuild/content/repo-projects/Directory.Build.props:
```diff
@@ -94,6 +94,17 @@
+    <TestActions>$(FlagParameterPrefix)restore</TestActions>
+    <TestActions>$(TestActions) $(FlagParameterPrefix)test</TestActions>
+
+    <TestArgs>$(FlagParameterPrefix)ci</TestArgs>
```
```suggestion
    <TestArgs Condition="'$(UseOfficialBuildVersioning)' != 'false'">$(FlagParameterPrefix)ci</TestArgs>
```

ViktorHofer commented on src/SourceBuild/content/repo-projects/installer.proj:
```diff
@@ -79,10 +79,12 @@
   <Target Name="CopySdkArchive"
           AfterTargets="Build"
           DependsOnTargets="DetermineSourceBuiltSdkVersion">
+
```
```suggestion
```

ViktorHofer commented on src/SourceBuild/content/repo-projects/installer.proj:
```diff
@@ -79,10 +79,12 @@
           DestinationFolder="$(ArtifactsAssetsDir)"
           SkipUnchangedFiles="true"
           UseHardlinksIfPossible="true" />
+
```
```suggestion
```

ViktorHofer commented on src/SourceBuild/content/test/tests.proj:
```diff
@@ -1,10 +1,29 @@
+  </PropertyGroup>
+
   <ItemGroup>
+    <ProjectReference Include="$(RepositoryEngineeringDir)extract-sdk-archive.proj" />
```
Move this ProjectReference into test/Directory.Build.props so that all projects depend on it by default. Declaring it here doesn't guarantee that it is available at build time of a test projects as the items here are built in parallel.

ViktorHofer commented on src/SourceBuild/content/test/scenario-tests/scenario-tests.proj:
```diff
@@ -0,0 +1,16 @@
+<Project Sdk="Microsoft.Build.NoTargets">
+  <PropertyGroup>
```
style nit:

```suggestion

  <PropertyGroup>
```

ViktorHofer commented on src/SourceBuild/content/test/scenario-tests/scenario-tests.proj:
```diff
@@ -0,0 +1,16 @@
+<Project Sdk="Microsoft.Build.NoTargets">
+  <PropertyGroup>
+    <TargetFramework>$(NetCurrent)</TargetFramework>
+    <IsTestProject>true</IsTestProject>
```
Just for my own understanding, why did you need to add this property?

ViktorHofer commented on src/SourceBuild/content/repo-projects/scenario-tests.proj:
```diff
@@ -10,4 +10,58 @@
+
+      <!-- It's necessary to explicitly define the path to the dotnet tool to prevent Arcade from attempting to derive it. Otherwise, it will run the dotnet
+           install script to get a new one. We must use the locally built SDK instead in order to support non-portable RIDs for source build. -->
+      <_DotNetTool Condition="'$(BuildOS)' == 'windows'">$(DotNetSdkExtractDir)dotnet.exe</_DotNetTool>
+      <_DotNetTool Condition="'$(BuildOS)' != 'windows'">$(DotNetSdkExtractDir)dotnet</_DotNetTool>
```
We usually just use the GetFileName function to avoid OS conditions:

```suggestion
      <_DotNetTool>$(DotNetSdkExtractDir)$([System.IO.Path]::GetFileName('$(DotNetTool)'))</_DotNetTool>
```

ViktorHofer commented on src/SourceBuild/content/repo-projects/scenario-tests.proj:
```diff
@@ -10,4 +10,58 @@
+    <ScenarioTestsArtifactsDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsDir)', 'scenario-tests'))</ScenarioTestsArtifactsDir>
+    <ScenarioTestsResultsDir>$([MSBuild]::NormalizeDirectory('$(ArtifactsTestResultsDir)', 'scenario-tests'))</ScenarioTestsResultsDir>
+    <_ScenarioTestsNuGetConfig>$(ScenarioTestsArtifactsDir)NuGet.config</_ScenarioTestsNuGetConfig>
+    <_InstallerNuGetConfig>$(RepoRoot)src/installer/NuGet.config</_InstallerNuGetConfig>
```
```suggestion
    <_InstallerNuGetConfig>$([MSBuild]::NormalizePath('$(SrcDir)', 'installer', 'NuGet.config'))</_InstallerNuGetConfig>
```

ViktorHofer commented on src/SourceBuild/content/test/tests.proj:
```diff
@@ -1,10 +1,29 @@
+    <!-- Skip scenario tests if the host architecture is different from the target architecture since the tests
+        require the ability to execute the built SDK. But the CLI is not capable of running on a host with a 
+        different architecture (i.e. "cannot execute binary file: Exec format error"). -->
+    <_RunScenarioTests Condition="'$(BuildArchitecture.ToLower())' != '$(TargetArchitecture.ToLower())'">false</_RunScenarioTests>
```
```suggestion
    <_RunScenarioTests Condition="'$(BuildArchitecture.ToLowerInvariant())' != '$(TargetArchitecture.ToLowerInvariant())'">false</_RunScenarioTests>
```

ViktorHofer commented on src/SourceBuild/content/test/tests.proj:
```diff
@@ -1,10 +1,29 @@
+    <!-- Skip scenario tests if the portable OS (determined from the host machine) is different from the target OS
+        since the tests require the ability to execute the built SDK. An example of where this would be disabled is
+        cross-build of using Mariner to build for Alpine (linux vs linux-musl). -->
+    <_RunScenarioTests Condition="'$(BuildOS)' != 'windows' and '$(__PortableTargetOS.ToLower())' != '$(TargetOS.ToLower())'">false</_RunScenarioTests>
```
```suggestion
    <_RunScenarioTests Condition="'$(BuildOS)' != 'windows' and '$(__PortableTargetOS.ToLowerInvariant())' != '$(TargetOS.ToLowerInvariant())'">false</_RunScenarioTests>
```

ViktorHofer commented on src/SourceBuild/content/test/tests.proj:
```diff
@@ -1,10 +1,29 @@
+        cross-build of using Mariner to build for Alpine (linux vs linux-musl). -->
+    <_RunScenarioTests Condition="'$(BuildOS)' != 'windows' and '$(__PortableTargetOS.ToLower())' != '$(TargetOS.ToLower())'">false</_RunScenarioTests>
+
+    <!-- The scenario tests are not supported when unofficial build versioning is used. -->
```
Please file a tracking issue for that and link to it here.

mthalman replied:
https://github.com/dotnet/source-build/issues/4364

mthalman commented:
A new error is showing up from the Aspire test. This isn't specific to my changes. I can repro it using the latest built SDK from the unified build pipeline. This seems specific to the unified build though. I can't repro it using an SDK built from the official installer pipeline. This error also doesn't occur in the source build pipeline from this PR.

```
Failed to instatiate template '.NET Aspire Application', the following constraints are not met:
  workload: The constraint 'workload' failed to initialize: Manifest provider Microsoft.NET.Sdk.WorkloadManifestReader.SdkDirectoryWorkloadManifestProvider returned a duplicate manifest ID 'microsoft.net.sdk.android' [/vmr/artifacts/obj/extracted-dotnet-sdk/sdk-manifests/9.0.100-preview.1/microsoft.net.sdk.android/34.99.0-preview.1.151/WorkloadManifest.json] that conflicts with existing manifest [/vmr/artifacts/obj/extracted-dotnet-sdk/sdk-manifests/9.0.100-preview.1/microsoft.net.sdk.android/34.99.0-preview.1.151/WorkloadManifest.json]
    
```

@Forgind - It looks like you work in the workload space. Can you help point where to look in diagnosing this error?

mthalman commented:
> A new error is showing up from the Aspire test. This isn't specific to my changes. I can repro it using the latest built SDK from the unified build pipeline. This seems specific to the unified build though. I can't repro it using an SDK built from the official installer pipeline. This error also doesn't occur in the source build pipeline from this PR.
> 
> ```
> Failed to instatiate template '.NET Aspire Application', the following constraints are not met:
>   workload: The constraint 'workload' failed to initialize: Manifest provider Microsoft.NET.Sdk.WorkloadManifestReader.SdkDirectoryWorkloadManifestProvider returned a duplicate manifest ID 'microsoft.net.sdk.android' [/vmr/artifacts/obj/extracted-dotnet-sdk/sdk-manifests/9.0.100-preview.1/microsoft.net.sdk.android/34.99.0-preview.1.151/WorkloadManifest.json] that conflicts with existing manifest [/vmr/artifacts/obj/extracted-dotnet-sdk/sdk-manifests/9.0.100-preview.1/microsoft.net.sdk.android/34.99.0-preview.1.151/WorkloadManifest.json]
>     
> ```
> 
> @Forgind - It looks like you work in the workload space. Can you help point where to look in diagnosing this error?
> 
> [Build link](https://dev.azure.com/dnceng-public/public/_build/results?buildId=656802&view=logs&jobId=651e3687-9c94-5c44-2c9d-36abed874f6c&j=651e3687-9c94-5c44-2c9d-36abed874f6c&t=76d4a79f-6e0f-5d18-0978-fead18ff8c43)

Now the latest build shows unified build passing but source build is failing with a similar error, but referring to a different package:

```
Failed to instatiate template '.NET Aspire Application', the following constraints are not met:
  workload: The constraint 'workload' failed to initialize: Manifest provider Microsoft.NET.Sdk.WorkloadManifestReader.SdkDirectoryWorkloadManifestProvider returned a duplicate manifest ID 'microsoft.net.workload.emscripten.net7' [/vmr/artifacts/obj/extracted-dotnet-sdk/sdk-manifests/9.0.100-preview.4/microsoft.net.workload.emscripten.net7/9.0.0-preview.4.24215.3/WorkloadManifest.json] that conflicts with existing manifest [/vmr/artifacts/obj/extracted-dotnet-sdk/sdk-manifests/9.0.100-preview.4/microsoft.net.workload.emscripten.net7/9.0.0-preview.4.24215.3/WorkloadManifest.json]
```

[Build link](https://dev.azure.com/dnceng-public/public/_build/results?buildId=657267&view=logs&j=609589e2-4f74-5576-cdb7-914bcaea778b&t=23fc9008-9397-50b6-064a-d79517ded190)

That would seem to imply the issue is intermittent.

Forgind commented:
> > A new error is showing up from the Aspire test. This isn't specific to my changes. I can repro it using the latest built SDK from the unified build pipeline. This seems specific to the unified build though. I can't repro it using an SDK built from the official installer pipeline. This error also doesn't occur in the source build pipeline from this PR.
> > ```
> > Failed to instatiate template '.NET Aspire Application', the following constraints are not met:
> >   workload: The constraint 'workload' failed to initialize: Manifest provider Microsoft.NET.Sdk.WorkloadManifestReader.SdkDirectoryWorkloadManifestProvider returned a duplicate manifest ID 'microsoft.net.sdk.android' [/vmr/artifacts/obj/extracted-dotnet-sdk/sdk-manifests/9.0.100-preview.1/microsoft.net.sdk.android/34.99.0-preview.1.151/WorkloadManifest.json] that conflicts with existing manifest [/vmr/artifacts/obj/extracted-dotnet-sdk/sdk-manifests/9.0.100-preview.1/microsoft.net.sdk.android/34.99.0-preview.1.151/WorkloadManifest.json]
> >     
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > @Forgind - It looks like you work in the workload space. Can you help point where to look in diagnosing this error?
> > [Build link](https://dev.azure.com/dnceng-public/public/_build/results?buildId=656802&view=logs&jobId=651e3687-9c94-5c44-2c9d-36abed874f6c&j=651e3687-9c94-5c44-2c9d-36abed874f6c&t=76d4a79f-6e0f-5d18-0978-fead18ff8c43)
> 
> Now the latest build shows unified build passing but source build is failing with a similar error, but referring to a different package:
> 
> ```
> Failed to instatiate template '.NET Aspire Application', the following constraints are not met:
>   workload: The constraint 'workload' failed to initialize: Manifest provider Microsoft.NET.Sdk.WorkloadManifestReader.SdkDirectoryWorkloadManifestProvider returned a duplicate manifest ID 'microsoft.net.workload.emscripten.net7' [/vmr/artifacts/obj/extracted-dotnet-sdk/sdk-manifests/9.0.100-preview.4/microsoft.net.workload.emscripten.net7/9.0.0-preview.4.24215.3/WorkloadManifest.json] that conflicts with existing manifest [/vmr/artifacts/obj/extracted-dotnet-sdk/sdk-manifests/9.0.100-preview.4/microsoft.net.workload.emscripten.net7/9.0.0-preview.4.24215.3/WorkloadManifest.json]
> ```
> 
> [Build link](https://dev.azure.com/dnceng-public/public/_build/results?buildId=657267&view=logs&j=609589e2-4f74-5576-cdb7-914bcaea778b&t=23fc9008-9397-50b6-064a-d79517ded190)
> 
> That would seem to imply the issue is intermittent.

That's interesting. The workload resolver supports there being multiple workload set files. If there are, it reads them all in and composes them, erroring if there are any duplicate entries because we clearly can't have two different versions of a single workload at the same time.

It appears that in this case, we're loading the same workload set twice, which means that there are definitely going to be duplicates—in fact, the whole file should be duplicates. So I see two bugs here: first, we should accept duplicates as long as they're the exact same version (basically drop the error in this case). Second, we shouldn't be loading the same manifest twice.

The first is an easy change. The second will take more investigation, as I don't know why we're trying to load the same manifest twice...glancing very quickly through the code, I don't see a way that's possible, but I didn't look exhaustively. Perhaps this has something to do with the overlay resolver? Perhaps something external is using reflection to call LoadManifestsFromProvider without updating _initializedManifests?

ViktorHofer commented on eng/pipelines/templates/jobs/vmr-build.yml:
```diff
@@ -455,6 +455,7 @@ jobs:
         cd "$(sourcesPath)"
         find artifacts/log/ -type f -name "*.binlog" -exec rsync -R {} -t ${targetFolder} \;
         find artifacts/log/ -type f -name "*.log" -exec rsync -R {} -t ${targetFolder} \;
+        [ -d "artifacts/scenario-tests/" ] && find artifacts/scenario-tests/ -type f -name "*.binlog" -exec rsync -R {} -t ${targetFolder} \;
```
What about the Windows counterpart?
